osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java

```diff
@@ -128,6 +277,289 @@ Date readClock(
+      // - No scalerUnit is defined or the scalerUnit is defined as Dynamic. In that case, the
+      //   scaler unit needs to be read from the meter
+      if (register.needsScalerUnitFromMeter()) {
+        captureObjectsThatNeedScalerUnitFromMeter.add(captureObject);
+      }
+    }
+
+    // Get scaler units from meter. They are read from the meter in one call for efficiency.
+    final Map<CaptureObject, String> scalerUnits =
+        new HashMap<>(
+            this.getScalerUnitsFromMeter(captureObjectsThatNeedScalerUnitFromMeter, conn, device));
+
+    // Create a new list with capture objects and fill in the missing scaler units.
+    // Note: the order should be the same as the order of the capture objects in the input param.
+    for (final CaptureObject captureObject : captureObjects) {
+      if (scalerUnits.containsKey(captureObject)) {
+        final CosemObject cosemObject = captureObject.getCosemObject();
+        final Attribute scalerUnitAttribute =
+            cosemObject.getAttribute(ExtendedRegisterAttribute.SCALER_UNIT.attributeId());
+
+        final Attribute newScalerUnitAttribute =
+            scalerUnitAttribute.copyWithNewValue(scalerUnits.get(captureObject));
+
+        captureObjectsWithScalerUnit.add(
+            captureObject.copyWithNewAttribute(newScalerUnitAttribute));
+      } else {
+        captureObjectsWithScalerUnit.add(captureObject);
+      }
+    }
+
+    return captureObjectsWithScalerUnit;
+  }
+
+  private List<CaptureObject> getRelevantCaptureObjects(
+      final List<CaptureObject> captureObjects, final String medium, final int channel) {
+    return captureObjects.stream()
+        .filter(captureObject -> captureObject.getCosemObject() instanceof Register)
+        .filter(
+            captureObject ->
+                captureObject.getAttributeId() == ExtendedRegisterAttribute.VALUE.attributeId())
+        .filter(captureObject -> captureObject.getCosemObject().getGroup().equals(medium))
+        .filter(
+            captureObject ->
+                this.getChannelWithoutException(captureObject.getCosemObject()) == channel)
+        .toList();
+  }
+
+  private int getChannelWithoutException(final CosemObject object) {
+    try {
+      return object.getChannel();
+    } catch (final ObjectConfigException e) {
+      return -1;
+    }
+  }
+
+  private Map<CaptureObject, String> getScalerUnitsFromMeter(
+      final List<CaptureObject> captureObjects,
+      final DlmsConnectionManager conn,
+      final DlmsDevice device)
+      throws ProtocolAdapterException {
+    if (captureObjects.isEmpty()) {
+      return Map.of();
+    }
+
+    final Map<CaptureObject, String> captureObjectsWithScalerUnit = new HashMap<>();
+
+    final AttributeAddress[] scalerUnitAddresses = this.getScalerUnitAddresses(captureObjects);
+
+    conn.getDlmsMessageListener()
+        .setDescription(
+            String.format(
+                "GetPeriodicMeterReadsGas scaler units, retrieve attribute: %s",

```

Comment:
This description could be more generic since this abstract class reads G and E devices